Failed to import projects because BasicFileDetector could not scan dirs

```
Failed to import projects
Failed to scan <REDACTED: user-file-path>
org.eclipse.core.runtime.CoreException: Failed to scan <REDACTED: user-file-path>
	at org.eclipse.jdt.ls.core.internal.managers.BasicFileDetector.scan(BasicFileDetector.java:140)
	at org.eclipse.jdt.ls.core.internal.managers.GradleProjectImporter.applies(GradleProjectImporter.java:154)
	at org.eclipse.jdt.ls.core.internal.managers.ProjectsManager.importProjects(ProjectsManager.java:151)
	at org.eclipse.jdt.ls.core.internal.managers.ProjectsManager.initializeProjects(ProjectsManager.java:114)
	at org.eclipse.jdt.ls.core.internal.handlers.InitHandler$1.runInWorkspace(InitHandler.java:260)
	at org.eclipse.core.internal.resources.InternalWorkspaceJob.run(InternalWorkspaceJob.java:43)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.nio.file.NoSuchFileException: <REDACTED: user-file-path>
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>$Basic.readAttributes(Unknown Source)
	at <REDACTED: user-file-path>$Basic.readAttributes(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at org.eclipse.jdt.ls.core.internal.managers.BasicFileDetector.scanDir(BasicFileDetector.java:178)
	at org.eclipse.jdt.ls.core.internal.managers.BasicFileDetector.scan(BasicFileDetector.java:138)
	... 6 more
```

```
Failed to import projects
Failed to scan <REDACTED: user-file-path>
org.eclipse.core.runtime.CoreException: Failed to scan <REDACTED: user-file-path>
	at org.eclipse.jdt.ls.core.internal.managers.BasicFileDetector.scan(BasicFileDetector.java:140)
	at org.eclipse.jdt.ls.core.internal.managers.GradleProjectImporter.applies(GradleProjectImporter.java:154)
	at org.eclipse.jdt.ls.core.internal.managers.ProjectsManager.importProjects(ProjectsManager.java:151)
	at org.eclipse.jdt.ls.core.internal.managers.ProjectsManager.initializeProjects(ProjectsManager.java:114)
	at org.eclipse.jdt.ls.core.internal.handlers.InitHandler$1.runInWorkspace(InitHandler.java:260)
	at org.eclipse.core.internal.resources.InternalWorkspaceJob.run(InternalWorkspaceJob.java:43)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.nio.file.AccessDeniedException: <REDACTED: user-file-path>
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at org.eclipse.jdt.ls.core.internal.managers.BasicFileDetector.scanDir(BasicFileDetector.java:178)
	at org.eclipse.jdt.ls.core.internal.managers.BasicFileDetector.scan(BasicFileDetector.java:138)
	... 6 more
```

```
Failed to import projects
Failed to scan <REDACTED: user-file-path>
org.eclipse.core.runtime.CoreException: Failed to scan <REDACTED: user-file-path>
	at org.eclipse.jdt.ls.core.internal.managers.BasicFileDetector.scan(BasicFileDetector.java:140)
	at org.eclipse.jdt.ls.core.internal.managers.GradleProjectImporter.applies(GradleProjectImporter.java:154)
	at org.eclipse.jdt.ls.core.internal.managers.ProjectsManager.importProjects(ProjectsManager.java:151)
	at org.eclipse.jdt.ls.core.internal.managers.ProjectsManager.initializeProjects(ProjectsManager.java:114)
	at org.eclipse.jdt.ls.core.internal.handlers.InitHandler$1.runInWorkspace(InitHandler.java:260)
	at org.eclipse.core.internal.resources.InternalWorkspaceJob.run(InternalWorkspaceJob.java:43)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.nio.file.FileSystemException: <REDACTED: user-file-path> <file>: Operation not permitted
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at <REDACTED: user-file-path>(Unknown Source)
	at org.eclipse.jdt.ls.core.internal.managers.BasicFileDetector.scanDir(BasicFileDetector.java:178)
	at org.eclipse.jdt.ls.core.internal.managers.BasicFileDetector.scan(BasicFileDetector.java:138)
	... 6 more

```

--- a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/BasicFileDetector.java
+++ b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/BasicFileDetector.java
@@ -17,12 +17,15 @@ import static java.nio.file.FileVisitResult.SKIP_SUBTREE;
 import static java.nio.file.FileVisitResult.TERMINATE;
 
 import java.io.IOException;
+import java.nio.file.AccessDeniedException;
+import java.nio.file.FileSystemException;
 import java.nio.file.FileSystemLoopException;
 import java.nio.file.FileSystems;
 import java.nio.file.FileVisitOption;
 import java.nio.file.FileVisitResult;
 import java.nio.file.FileVisitor;
 import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
 import java.nio.file.Path;
 import java.nio.file.PathMatcher;
 import java.nio.file.SimpleFileVisitor;
@@ -168,6 +171,12 @@ public class BasicFileDetector {
 			public FileVisitResult visitFileFailed(Path file, IOException exc) throws IOException {
 				Objects.requireNonNull(file);
 				if (exc instanceof FileSystemLoopException) {
+					return CONTINUE;
+				} else if (exc instanceof NoSuchFileException || exc instanceof AccessDeniedException
+					|| exc instanceof FileSystemException) {
+					// Permission or file not found issue is more likely a user error,
+					// we should skip it and continue.
+					JavaLanguageServerPlugin.logInfo("Scan the file failed: " + exc.toString());
 					return CONTINUE;
 				} else {
 					throw exc;

--- a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/BasicFileDetector.java
+++ b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/BasicFileDetector.java
@@ -176,7 +176,7 @@ public class BasicFileDetector {
 					|| exc instanceof FileSystemException) {
 					// Permission or file not found issue is more likely a user error,
 					// we should skip it and continue.
-					JavaLanguageServerPlugin.logInfo("Scan the file failed: " + exc.toString());
+					JavaLanguageServerPlugin.logInfo("Scan of file failed: " + exc.toString());
 					return CONTINUE;
 				} else {
 					throw exc;

--- a/org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/managers/BasicFileDetectorTest.java
+++ b/org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/managers/BasicFileDetectorTest.java
@@ -234,6 +234,15 @@ public class BasicFileDetectorTest {
 		}
 	}
 
+	@Test
+	public void testScanNotFoundDirectory() throws Exception {
+		File notFoundDirectory = new File(System.getProperty("java.io.tmpdir"), "foo_bar_not_found_" +
+			+ new Random().nextInt(10000));
+		BasicFileDetector detector = new BasicFileDetector(notFoundDirectory.toPath(), "buildfile");
+		Collection<Path> dirs = detector.scan(null);
+		assertEquals("Found " + dirs, 0, dirs.size()); // No uncaught exception occurs
+	}
+
 	@SafeVarargs
 	private final <E> List<E> list(E... elements) {
 		return new ArrayList<>(Arrays.asList(elements));

